sdk/iot/hub/samples/telemetry/src/iot_hub_telemetry_example.c

```diff
@@ -0,0 +1,278 @@
+      NULL));
+
+  return AZ_OK;
+}
+
+static int connect_device()
+{
+  int rc;
+
+  MQTTClient_SSLOptions mqtt_ssl_options = MQTTClient_SSLOptions_initializer;
+  MQTTClient_connectOptions mqtt_connect_options = MQTTClient_connectOptions_initializer;
+
+  char username[128];
+  if ((rc = az_iot_hub_client_get_user_name(&client, username, sizeof(username), NULL)) != AZ_OK)
+
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  mqtt_connect_options.username = username;
+  mqtt_connect_options.password = NULL; // Using X509 Client Certificate authentication.
+
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  if (*x509_trust_pem_file != '\0')
+  {
+    mqtt_ssl_options.trustStore = (char*)x509_trust_pem_file;
+  }
+
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to connect, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int send_telemetry_messages()
+{
+  int rc;
+  for (int i = 0; i < NUMBER_OF_MESSAGES; ++i)
+  {
+    printf("Sending Message %i\n", i + 1);
+    if ((rc = MQTTClient_publish(
+             mqtt_client,
+             telemetry_topic,
+             (int)strlen(telemetry_message_payloads[i]),
+             telemetry_message_payloads[i],
+             0,
+             0,
+             NULL))
+        != MQTTCLIENT_SUCCESS)
+    {
+      printf("Failed to publish telemetry topic, return code %d\n", rc);
+      return rc;
+    }
+    sleep_seconds(1);
+  }
+  return rc;
+}
+
+int main()
+{
+  int rc;
+
+  if ((rc = read_configuration_and_init_client()) != AZ_OK)
+  {
+    printf("Failed to read configuration from environment variables, return code %d\n", rc);
+    return rc;
+  }
+
+  char client_id[128];
+  size_t client_id_length;
+  if ((rc
+       = az_iot_hub_client_get_client_id(&client, client_id, sizeof(client_id), &client_id_length))
+      != AZ_OK)
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_create(
+           &mqtt_client, mqtt_endpoint, client_id, MQTTCLIENT_PERSISTENCE_NONE, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to create MQTT client, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = connect_device()) != 0)
+  {
+    return rc;
+  }
+
+  if ((rc = az_iot_hub_client_telemetry_get_publish_topic(
+           &client, NULL, telemetry_topic, sizeof(telemetry_topic), NULL))
+      != AZ_OK)
+  {
+    return rc;
+  }
+
+  // Loop and send 5 messages
+  if ((rc = send_telemetry_messages()) != 0)
+  {
+    return rc;
+  }
+
+  printf("Messages Sent [Press ENTER to shut down]\n");
+  (void)getchar();
+
+  if ((rc = MQTTClient_disconnect(mqtt_client, 10000)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to disconnect MQTT client, return code %d\n", rc);
+    return rc;
```

Comment:
Is there a reason we are not following clean shut down approach in failure case as we did in sdkv2
[https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_ll_telemetry_sample/iothub_ll_telemetry_sample.c](https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_ll_telemetry_sample/iothub_ll_telemetry_sample.c)

Reply:
I left comments in the provisioning sample: apps should crash immediately unless it can handle the error code.
If the error is out of memory or an AV continuing would destroy the state and make the bug impossible to debug.

It's more or less like catching Exception in java/C# which is wrong: https://softwareengineering.stackexchange.com/questions/270406/is-it-a-good-practice-to-catch-all-exception

Reply:
Every failure can't be categorized as unhandled failure and state of application should be closed properly if we are able to handle failure. By abruptly closing the application you are not playing well with our service.
e.g by not sending explicit RST packet because you were not able to full file some application logic is not playing well with service and is causing resource hogging on service side. As this example come from Azure, we should be mindful of this.  

---
In reply to: [418697016](https://github.com/Azure/azure-sdk-for-c/pull/648#discussion_r418697016) [](ancestors = 418697016)